test/Plutip/UtxoDistribution.purs

```diff
@@ -0,0 +1,214 @@
+module Test.Plutip.UtxoDistribution
+  ( ArbitraryUtxoDistr
+  , assertContract
+  , assertCorrectDistribution
+  , assertNoUtxosAtAddress
+  , assertNoUtxosAtEnterpriseAddress
+  , assertUtxosAtPlutipWalletAddress
+  , checkUtxoDistribution
+  , genInitialUtxo
+  , suite
+  , withArbUtxoDistr
+  ) where
+
+import Prelude
+
+import Contract.Address
+  ( Address
+  , getNetworkId
+  , getWalletAddress
+  , ownPaymentPubKeyHash
+  , ownStakePubKeyHash
+  , payPubKeyHashEnterpriseAddress
+  )
+import Contract.Monad (Contract, liftedM)
+import Contract.Test.Plutip
+  ( class UtxoDistribution
+  , InitialUTxOs
+  , runPlutipContract
+  )
+import Contract.Transaction
+  ( TransactionInput
+  , TransactionOutput(TransactionOutput)
+  )
+import Contract.Utxos (utxosAt)
+import Contract.Value (Value, lovelaceValueOf)
+import Contract.Wallet (KeyWallet, withKeyWallet)
+import Control.Lazy (fix)
+import Data.Array (foldl, zip)
+import Data.BigInt (BigInt)
+import Data.BigInt as BigInt
+import Data.FoldableWithIndex (foldlWithIndex)
+import Data.List as List
+import Data.Map as Map
+import Data.Maybe (isJust)
+import Data.Newtype (unwrap, wrap)
+import Data.NonEmpty ((:|))
+import Data.Traversable (for_)
+import Data.Tuple.Nested (type (/\), (/\))
+import Effect.Class (liftEffect)
+import Effect.Exception (throw)
+import Mote (group, test)
+import Plutip.Types (InitialUTxOsWithStakeKey(InitialUTxOsWithStakeKey))
+import Plutip.UtxoDistribution (encodeDistribution, keyWallets)
+import Plutus.Types.Transaction (Utxo)
+import Test.Plutip.Common (config, privateStakeKey)
+import Test.QuickCheck (class Arbitrary, arbitrary)
+import Test.QuickCheck.Gen (Gen, arrayOf, chooseInt, frequency, randomSample')
+import TestM (TestPlanM)
+import Type.Prelude (Proxy(Proxy))
+
+suite :: TestPlanM Unit
+suite = group "Plutip UtxoDistribution" do
+  distrs <- liftEffect $ randomSample' 5 arbitrary
+  for_ distrs $ \distr ->
+    test
+      ( "runPlutipContract: stake key transfers with random distribution: "
+          <> show distr
+      )
+      $
+        withArbUtxoDistr
+          distr
+          \randDistr -> runPlutipContract config randDistr $
+            checkUtxoDistribution randDistr
+
+checkUtxoDistribution
+  :: forall distr wallet (r :: Row Type)
+   . UtxoDistribution distr wallet
+  => distr
+  -> wallet
+  -> Contract r Unit
+checkUtxoDistribution distr wallets = do
+  let
+    walletsArray = keyWallets (Proxy :: Proxy distr) wallets
+    walletUtxos = encodeDistribution distr
+  for_ walletsArray assertUtxosAtPlutipWalletAddress
+  assertCorrectDistribution $ zip walletsArray walletUtxos
+
+-- TODO: minimum value of 1 ada is hardcoded, tests become flaky below
```

Comment:
We have a policy that each TODO should link to an open issue, Could you please add a link to https://github.com/Plutonomicon/cardano-transaction-lib/issues/857 in the comments?